projects/plugins/jetpack/_inc/client/components/popover/index.jsx

```diff
@@ -78,10 +81,10 @@ class Popover extends Component {
 
 	UNSAFE_componentWillReceiveProps( nextProps ) {
 		// update context (target) reference into a property
-		if ( ! isDOMElement( nextProps.context ) ) {
-			this.domContext = ReactDom.findDOMNode( nextProps.context );
+		if ( ! isDOMElement( nextProps.context ) && nextProps.context !== null ) {
+			this.domContextRef.current = nextProps.context.current;
```

Comment:
Is this going to work? Or is it likely that `nextProps.context.current` is undefined because `nextProps.context` is some arbitrary React element without a `.current` property?

Possibly we want to do something more like this, if being able to pass a ref is the intention
```js
if ( isDomNode( nextProps.context ) ) {
    this.domContextRef.current = nextProps.context;
} else if ( isRef( nextProps.context ) ) {
    this.domContextRef.current = nextProps.context.current;
    // Or should it be like this so changes to the ref make it in here? I don't really know React.
    this.domContextRef = nextProps.context;
} else if ( nextProps.context !== null ) {
    console.error( 'Expected a DOM node or a React ref', nextProps.context );
}
```

Reply:
Thank you, yes you're right - the new code assumes that if nextProps.context is not a DOM element, that it is a ref, but it may be neither. I also think the console.error is a good idea, as previously we were logging an error via isDomElement (isElement) if something was not an HTML element, but we need to check for refs here and also want to show an error otherwise so a custom solution is better.

>   Or should it be like this so changes to the ref make it in here? I don't really know React.
Oh , good question. From my testing it doesn't appear to make a difference.

From what I understand, it should be` this.domContextRef.current = nextProps.context.current` because assigning `this.domContextRef` directly to `nextProps.context` is overwriting the `this.domContextRef` reference, not just it's current property.

Reply:
I don't really know how React works, but my thinking was that if someone passes in a ref then they'd probably expect that changes to the ref affect the popover, rather than it using the value of the ref at initialization and then not checking it again when it's time to use it. But maybe React already handles that somehow?

Reply:
That's interesting, good point. Still learning and trying to understand this better too myself. And I realize I'm not thinking about cases when the ref itself would actually change. 

When the context prop changes, the `componentDidUpdate` function should run, and it should be possible to use that to update` this.domContextRef.current`.

Given that componentDidUpdate is [what is called after re-rendering due to a change in props or state](https://reactjs.org/docs/react-component.html#componentdidupdate) - and also given that `UNSAFE_componentWillReceiveProps` is legacy, adding this logic to `componentDidUpdate` should help make this a little more future proof too - and we can assign to .current in that case. And I believe we can remove `UNSAFE_componentWillReceiveProps` too with the snippet added to `componentDidUpdate`.

`setDOMBehaviour` looks like it only runs on construction so I've left that.

That all said and done, I wonder if it's worth bringing an internal React expert in to double check this perhaps 🤔

Reply:
> That all said and done, I wonder if it's worth bringing an internal React expert in to double check this perhaps 🤔

Sounds like a good plan to me!

Or what are the callers of this? Could we have the callers pass `whateverRef.current` so Popover only takes DOM nodes?

Reply:
> Or what are the callers of this? Could we have the callers pass whateverRef.current so Popover only takes DOM nodes?

Good point.

So as it turns out, all three callers of Popover already pass whateverRef.current. `ModuleChartBar`, used on the at-a-glance stats page, passes context as `this.valueBarRef.current` already. The `FormToggle` passes context as `this.toggleSwitchRef.current`. The `InfoPopover` passes context as `this.infoPopoverRef.curren`t.

The expected behaviour then is to pass in a DOM node, not a ref. So there is no need to create a workaround for refs, and it'd also be good to note that we expect DOM nodes too (perhaps that's clear with the new console.error messages but I added to some other comments with more context).

What I've done now to be safe is go back to using `UNSAFE_componentWillReceiveProps`, and I'll also see if I can 
 find someone with more React knowledge can confirm it really is safe to move the checks from `UNSAFE_componentWillReceiveProps` into `componentDidUpdate`, otherwise will keep it where it now is.